[TouchableOpacity] Reset opacity to the inactiveValue rather than always 1.0

brentvatne opened the pull request:
As per #941 - fixes bug with `TouchabeOpacity` always reseting child opacity to `1.0` after press.

A note about the code: we could probably use a general `getNativeProp(propName, callback)` function rather than `getOpacity` but just used that as it was simpler for this specific PR, perhaps that refactor could be left to another - or maybe there is a way to do this already that I missed.

Before:
![bug](https://cloud.githubusercontent.com/assets/90494/7287207/52d6a686-e907-11e4-8e16-04b2ddd0582c.gif)

After:
![after](https://cloud.githubusercontent.com/assets/90494/7287689/5aca4776-e90c-11e4-8c40-aa6bd3e822d8.gif)

Example code:
```javascript
'use strict';

var React = require('react-native');
var {
  AppRegistry,
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
} = React;


var TestIt = React.createClass({
  render() {
    return (
      <View style={styles.container}>
        <TouchableOpacity activeOpacity={0.3}>
          <View style={styles.searchButton}>
            <Text>Search</Text>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
});

var styles = StyleSheet.create({
  searchButton: {
    opacity: 0.5,
    width: 120,
    height: 50,
    top: 35,
    alignSelf: 'center',
    borderRadius: 10,
    borderColor: '#da766b',
    backgroundColor: '#e01a3c',
    borderWidth: 1,
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
});

AppRegistry.registerComponent('TestIt', () => TestIt);
```

tadeuzagallo commented:
What about:

```js
  touchableHandleActivePressOut: function() {
    var childStyle = this.refs[CHILD_REF].props.style;
    if (typeof childStyle === 'number') {
      childStyle = StyleSheetRegistry.getStyleByID(childStyle);
    }
    this.setOpacityTo(childStyle && childStyle.opacity || 1);
    this.props.onPressOut && this.props.onPressOut();
  },
```

And I don't think the opacity should be set on the `onPress` method (which already is).

/cc @vjeux

brentvatne commented:
@tadeuzagallo - I like that, I didn't realize we could just grab the opacity out of the registry like that - nifty!

brentvatne commented:
@tadeuzagallo - tested your code, works great, updated the PR (force-pushed so goodbye previous commit) to include it and removed setting opacity in `onPress` - does seem pointless given that it is set `onPressIn`

tadeuzagallo commented:
Ok, that looks good to me. I haven't actually seen `StyleSheetRegistry` being used before, so lets just see if @vjeux is ok with that

vjeux commented on Libraries/Components/Touchable/TouchableOpacity.js:
```diff
@@ -105,12 +106,17 @@ var TouchableOpacity = React.createClass({
+      childStyle = StyleSheetRegistry.getStyleByID(childStyle);
+    }
+
+    this.setOpacityTo(childStyle && childStyle.opacity || 1);
```
You cannot use || here. Otherwise if opacity is set to 0, it's going to be 1. You need childStyle.opacity === undefined ? 1 : childStyle.opacity

brentvatne replied:
Good point, that falsey 0

tadeuzagallo replied:
JavaScript :heart:

vjeux commented on Libraries/Components/Touchable/TouchableOpacity.js:
```diff
@@ -105,12 +106,17 @@ var TouchableOpacity = React.createClass({
-    this.setOpacityTo(1.0);
+    var childStyle = this.refs[CHILD_REF].props.style;
+
+    if (typeof childStyle === 'number') {
```
You want to use `flattenStyle` here. The type of childStyle can be a number, or an object, or an array of number or objects or array... :)

brentvatne commented:
@vjeux - thanks for the review, fixed

vjeux commented on Libraries/Components/Touchable/TouchableOpacity.js:
```diff
@@ -105,12 +107,12 @@ var TouchableOpacity = React.createClass({
 
   touchableHandleActivePressOut: function() {
-    this.setOpacityTo(1.0);
+    var childStyle = flattenStyle(this.refs[CHILD_REF].props.style || {});
```
can you put the || {} outside of the flattenStyle call? flattenStyle breaks early if it's undefined

vjeux commented on Libraries/Components/Touchable/TouchableOpacity.js:
```diff
@@ -14,6 +14,7 @@
 
 var NativeMethodsMixin = require('NativeMethodsMixin');
 var POPAnimationMixin = require('POPAnimationMixin');
+var StyleSheetRegistry = require('StyleSheetRegistry');
```
remove

vjeux commented on Libraries/Components/Touchable/TouchableOpacity.js:
```diff
@@ -22,6 +23,7 @@ var cloneWithProps = require('cloneWithProps');
 var ensureComponentIsNative = require('ensureComponentIsNative');
 var keyOf = require('keyOf');
 var onlyChild = require('onlyChild');
+var flattenStyle = require('flattenStyle');
```
nit: can you sort requires alphabetically?

brentvatne commented:
@vjeux - thanks for second round of comments, this PR gets smaller each time :stuck_out_tongue:

vjeux commented:
lgtm!

jessepollak commented:
+1 for this, thanks!
